src/utils.py

```diff
@@ -0,0 +1,85 @@
+"""Utils"""
+import os
+from pathlib import Path
+from typing import Set, Union
+
+import pandas as pd
+import yaml
+
+
+def add_prefixes_to_plain_id(x: str) -> str:
+    """From plain IDs from originanl source, add prefixes.
+
+    Terms:
+        CN: stands for "CUI Novel". These are created for any MedGen records without UMLS CUI.
+        C: stands for "CUI". These are sourced from UMLS.
+        CUI: stands for "Concept Unique Identifier"
+        UID (Unique IDentifier): These are cases where the id is all digits; does not start with a leading alpha char.
+    """
+    return f'MEDGENCUI:{x}' if x.startswith('CN') \
+        else f'UMLS:{x}' if x.startswith('C') \
+        else f'MEDGEN:{x}'
+
+
+def find_prefixes_in_mapping_set(source_df: pd.DataFrame) -> Set[str]:
+    """Find prefixes in mapping set"""
+    df = source_df.copy()
+    cols_with_prefixes = ['subject_id', 'object_id', 'predicate_id']
+    prefixes = set()
+    for col in cols_with_prefixes:
+        col2 = col.replace('id', 'prefix')
+        df[col2] = df[col].apply(lambda x: x.split(':')[0]
+            if isinstance(x, str) else x) # handles nan
+        prefixes.update(set(df[col2].to_list()))
+    return prefixes
+
+
+def write_sssom(df: pd.DataFrame, config_path: Union[Path, str], outpath: Union[Path, str]):
+    """Writes a SSSOM file with commented metadata at the top of the file.
+
+    Filters only prefxes in curie_map that exist in the mapping set."""
+    temp_filtered_config_path = str(config_path) + '.tmp'
+    # Load config
+    config = yaml.safe_load(open(config_path, 'r'))
+    # TODO: filter by prefixes in df
+    #  - filter them finally
+    # Filter curie_map
+    prefixes: Set[str] = find_prefixes_in_mapping_set(df)
+    config['curie_map'] = {k: v for k, v in config['curie_map'].items() if k in prefixes}
+    # Write
+    with open(temp_filtered_config_path, 'w') as f:
+        yaml.dump(config, f)
+    write_tsv_with_comments(df, temp_filtered_config_path, outpath)
+    os.remove(temp_filtered_config_path)
+
+
+def write_tsv_with_comments(df: pd.DataFrame, comments_file: Union[Path, str], outpath: Union[Path, str]):
```

Comment:
## General: `write_tsv_with_comments()`
@matentzn Let me know what the existing best practice is supposed to be for this.
This is a function I have in GARD, a TIMS repo, and possibly elsewhere. Maybe there's a function in `sssom-py` that does this like I want? I have a dataframe and a path to a `sssom-metadata.yml`, and I want to write that to a file and include the comment at the top.

Reply:
yep, 

1. instantiate a MappingSetDataFrame() object with the dataframe
2. use `write_table()` method to frite the result 

See example: https://github.com/mapping-commons/sssom-py/blob/be8a77b5a2ef42cd2b92b74d892d7626dec27471/src/sssom/cli.py#L280

Reply:
I recall this a bit now, thanks.

Reply:
Done~

Now my `write_sssom()` function has been rewritten to utilize `sssom-py`:
```python
def write_sssom(df: pd.DataFrame, config_path: Union[Path, str], outpath: Union[Path, str]):
    """Writes a SSSOM file"""
    with open(config_path, 'r') as yaml_file:
        metadata: Dict = yaml.load(yaml_file, Loader=yaml.FullLoader)
    converter = curies.Converter.from_prefix_map(metadata['curie_map'])
    msdf: MappingSetDataFrame = MappingSetDataFrame(converter=converter, df=df, metadata=metadata)
    with open(outpath, 'w') as f:
        write_table(msdf, f)
```

As with my [icd11 PR](https://github.com/monarch-initiative/icd11/pull/12#discussion_r1545497668) which had the same refactor, while working on this I also encountered some possible bugs and UX issues, and made:
- https://github.com/mapping-commons/sssom-py/issues/513
- https://github.com/mapping-commons/sssom-py/issues/514
- https://github.com/mapping-commons/sssom-py/issues/515
- https://github.com/mapping-commons/sssom-py/issues/516
- https://github.com/mapping-commons/sssom-py/issues/517